src/NServiceBus.Transport.SQS/InputQueuePump.cs

```diff
@@ -308,69 +317,111 @@ internal async Task ProcessMessage(Message receivedMessage, CancellationToken me
                 {
-                    immediateProcessingAttempts++;
-                    var errorHandlerResult = ErrorHandleResult.RetryRequired;
+                    criticalErrorAction(
+                        $"Failed to execute recoverability policy for message with native ID: `{nativeMessageId}`",
+                        onErrorEx,
+                        messageProcessingCancellationToken);
 
-                    try
-                    {
-                        errorHandlerResult = await onError(new ErrorContext(ex,
-                            new Dictionary<string, string>(headers),
-                            nativeMessageId,
-                            body,
-                            transportTransaction,
-                            immediateProcessingAttempts,
-                            ReceiveAddress,
-                            context), messageProcessingCancellationToken).ConfigureAwait(false);
-                    }
-                    catch (Exception onErrorEx) when (!onErrorEx.IsCausedBy(messageProcessingCancellationToken))
-                    {
-                        criticalErrorAction($"Failed to execute recoverability policy for message with native ID: `{nativeMessageId}`", onErrorEx, messageProcessingCancellationToken);
-                    }
+                    await ReturnMessageToQueue(nativeMessage).ConfigureAwait(false);
 
-                    errorHandled = errorHandlerResult == ErrorHandleResult.Handled;
+                    return;
                 }
             }
+
+            await DeleteMessage(nativeMessage, s3BodyKey, messageProcessingCancellationToken).ConfigureAwait(false);
+        }
+
+
+#pragma warning disable PS0018 //Cancellation token intentionally not passed because cancellation shouldn't stop messages from being returned to the queue
+        async Task ReturnMessageToQueue(Message message)
+#pragma warning restore PS0018
+        {
+            try
+            {
+                await sqsClient.ChangeMessageVisibilityAsync(
+                    new ChangeMessageVisibilityRequest()
+                    {
+                        QueueUrl = inputQueueUrl,
+                        ReceiptHandle = message.ReceiptHandle,
+                        VisibilityTimeout = 0
+                    }).ConfigureAwait(false);
+            }
+            catch (ReceiptHandleIsInvalidException ex)
+            {
+                Logger.Warn("Failed to return message to the queue. Receipt handle is invalid.", ex);
+            }
+            catch (MessageNotInflightException ex)
+            {
+                Logger.Warn("Failed to return message to the queue. The message is not in flight.", ex);
+            }
+            catch (Exception ex)
+            {
+                Logger.Warn(
+                    "Failed to return message to the queue. Message will return to queue after visibility expires.", ex);
+            }
+        }
+
+        int GetDeliveryAttempts(string messageIdKey)
```

Comment:
Let's rename `messageIdKey` to `nativeMessageId` to make it clearer what needs to be passed in.